src/Stratis.Bitcoin.Features.Wallet.Tests/WalletControllerTest.cs

```diff
@@ -346,6 +349,49 @@ public void RecoverWalletSuccessfullyReturnsWalletModel()
             Assert.Equal(200, viewResult.StatusCode);
         }
 
+        /// <summary>
+        /// This is to cover the scenario where a wallet is syncing at height X
+        /// and the user recovers a new wallet at height X + Y.
+        /// The wallet should continue syncing from X without jumpoing forward.
+        /// </summary>
+        [Fact]
+        public void RecoverWalletWithDatedAfterCurrentSyncHeightDoesNotMoveSyncHeight()
+        {
+            var wallet = new Wallet
+            {
+                Name = "myWallet",
+                Network = NetworkHelpers.GetNetwork("mainnet")
+            };
+
+            // The chain is at height 100.
+            ConcurrentChain concurrentChain = WalletTestsHelpers.GenerateChainWithHeight(100, this.Network);
+            DateTime lastBlockDateTime = concurrentChain.Tip.Header.BlockTime.DateTime;
+
+            var mockWalletWrapper = new Mock<IWalletManager>();
+            mockWalletWrapper.Setup(w => w.RecoverWallet(It.IsAny<string>(), It.IsAny<string>(), It.IsAny<string>(), It.IsAny<DateTime>(), null)).Returns(wallet);
+
+            Mock<IWalletSyncManager> walletSyncManager = new Mock<IWalletSyncManager>();
+            walletSyncManager.Setup(w => w.WalletTip).Returns(new ChainedHeader(this.Network.GetGenesis().Header, this.Network.GetGenesis().Header.GetHash(), 3));
+            walletSyncManager.Verify(w => w.SyncFromHeight(100), Times.Never);
+
+            var controller = new WalletController(this.LoggerFactory.Object, mockWalletWrapper.Object, new Mock<IWalletTransactionHandler>().Object, walletSyncManager.Object, It.IsAny<ConnectionManager>(), this.Network, this.chain, new Mock<IBroadcasterManager>().Object, DateTimeProvider.Default);
+
+            IActionResult result = controller.Recover(new WalletRecoveryRequest
+            {
+                Name = "myWallet",
+                FolderPath = "",
+                Password = "",
+                Network = "MainNet",
+                Mnemonic = "mnemonic",
+                CreationDate = lastBlockDateTime
+            });
+
+            mockWalletWrapper.VerifyAll();
```

Comment:
Is this necessary? the test wouldn't work if the setups didn't work.

Also stacking up asserts in the arrange, to verify in the assert blurs the boundaries between arrange act and assert.   If an explicit Verify is needed (meaning there is no side effect to test other than knowing that method got called)  it can happen explicity in the Assert section (here)

Reply:
I guess calling VerifyAll is more for making sure everything has been verified.
We should look into making the difference between arrange and assert more distinct, but let's leave it for another time. I'd like to do it for all the tests at some point.